mojo/proposals/drain-collection-support.md

```diff
@@ -0,0 +1,452 @@
+
+## Proposal
+
+The stdlib team has proposed to move forward with [Rust style draining of
+collections][5].
+
+Scope:
+
+This proposal can expand to support additional
+collections beyond `List`, `Set`, and `Dict`, however for brevity we will focus
+on the basic collections for now. It also goes through compatibility with iter
+tools like `zip`, however will not go into compatibility of all the iter tool
+functions that could be added in the future that Python supports.
+
+Goals:
+
+- `take_items` implementation for `Set`, `List`, and `Dict` initially that aligns
+with Rust's `Drain` struct that does not copy elements.
+- Compatibility with iter tools like `zip`.
+
+Non-Goals:
+
+- Block on move-only tuple unpacking.
+
+### User Experience Cases
+
+In all cases, `ref` must be used to get mutable access to the item. `for` loops
+by default iterate over immutable references (even if the source collection is
+mutable).
+
+In all cases the call to `take_items` takes mutable reference of the underlying
+collection, and clears the source collection immediately on destruction of the
+iterator. Cases such as `take_items(Slice(..))` takes mutable reference of the
+subset of the collection, and updates the source collection on destruction of
+the iterator.
+
+- e.g.: `list.take_items(Slice(1, 2))` takes elements at indicies 1 and 2, and
+one-time shifts left the remaining elements in the source list on destruction of
+the iterator.
+
+Basic flow from Collection -> TakeIterable -> for loop / iteration as
+pseudo-code.
+
+```mojo
+collection = List("a", "b", "c") or Dict("a": 1, "b": 2, "c": 3) or \ 
+Set(1, 2, 3)
+take_iter = collection.take_items()
+assert(len(collection) == 3) 
+
+for ref item in take_iter:
+    # Do something with item or break early
+    # Note: What happens if we do `collection.pop(0)` here? Is this something 
+    # to worry about?
+    break
+# On take_iter destruction, the collection (or its subset) is cleared / 
+# updated.
+assert(len(collection) == 0) 
+```
+
+A non-obvious case here is how to handle cases where iterated items are stored
+in tuples such as in `zip`. The examples below, wrap the elements in `Optional`
+so we can `take` the elements out. This allows the elements to be moved out of
+tuples without deinitializing the tuple overall. As part of this proposal a
+method such as `DictEntry.reap` will return a
+`Tuple[Optional[Key], Optional[Value]]` that allows each element to be
+independently moved out of the tuple without deinitializing the tuple overall.
+
+- Ideally above would not be needed if structs (with a unpackable trait) and
+tuples supported move unpacking. Reference issue
+[#5330](https://github.com/modular/modular/issues/5330).
+
+The basic case is a complete drain of each respective collection.
+
+```mojo
+def basic_user_experience_cases():
+    print('Testing list')
+    var l = List[Foo](Foo(x="a"), Foo(x="b"), Foo(x="c"))
+    for ref item in l.take_items():
```

Comment:
I don't believe you want `ref` here. Since `take_items` returns "owned" values that is it draining from the collection.


```suggestion
    for item in l.take_items():
```